app/scripts/metamask-controller.js

```diff
@@ -478,1922 +451,2104 @@ export default class MetamaskController extends EventEmitter {
-            ],
-          }),
+        disabled: !this.preferencesController.state.useExternalServices,
+        clientConfigApiService: new ClientConfigApiService({
+          fetch: globalThis.fetch.bind(globalThis),
+          config: {
+            client: ClientType.Extension,
+            distribution:
+              this._getConfigForRemoteFeatureFlagRequest().distribution,
+            environment:
+              this._getConfigForRemoteFeatureFlagRequest().environment,
+          },
         }),
-      ],
-      state: initState.NameController,
-    });
+      });
 
-    const petnamesBridgeMessenger = this.controllerMessenger.getRestricted({
-      name: 'PetnamesBridge',
-      allowedEvents: [
-        'NameController:stateChange',
-        'AccountsController:stateChange',
-        'AddressBookController:stateChange',
-      ],
-      allowedActions: ['AccountsController:listAccounts'],
-    });
+      // RemoteFeatureFlagController has subscription for preferences changes
+      this.controllerMessenger.subscribe(
+        'PreferencesController:stateChange',
+        previousValueComparator((prevState, currState) => {
+          const { useExternalServices: prevUseExternalServices } = prevState;
+          const { useExternalServices: currUseExternalServices } = currState;
+          if (currUseExternalServices && !prevUseExternalServices) {
+            this.remoteFeatureFlagController.enable();
+            this.remoteFeatureFlagController.updateRemoteFeatureFlags();
+          } else if (!currUseExternalServices && prevUseExternalServices) {
+            this.remoteFeatureFlagController.disable();
+          }
+        }, this.preferencesController.state),
+      );
 
-    new AddressBookPetnamesBridge({
-      addressBookController: this.addressBookController,
-      nameController: this.nameController,
-      messenger: petnamesBridgeMessenger,
-    }).init();
-
-    new AccountIdentitiesPetnamesBridge({
-      nameController: this.nameController,
-      messenger: petnamesBridgeMessenger,
-    }).init();
-
-    this.userOperationController = new UserOperationController({
-      entrypoint: process.env.EIP_4337_ENTRYPOINT,
-      getGasFeeEstimates: this.gasFeeController.fetchGasFeeEstimates.bind(
-        this.gasFeeController,
-      ),
-      messenger: this.controllerMessenger.getRestricted({
-        name: 'UserOperationController',
-        allowedActions: [
-          'ApprovalController:addRequest',
-          'NetworkController:getNetworkClientById',
-          'KeyringController:prepareUserOperation',
-          'KeyringController:patchUserOperation',
-          'KeyringController:signUserOperation',
-        ],
-      }),
-      state: initState.UserOperationController,
-    });
+      return this.remoteFeatureFlagController;
+    };
 
-    this.userOperationController.hub.on(
-      'user-operation-added',
-      this._onUserOperationAdded.bind(this),
-    );
+    const controllerInitObjects = [
```

Comment:
In future, we could automatically order this list via each init class explicitly declaring it's dependencies with a new abstract method such as `getControllerDependencies`.

Alternatively, the order may not matter if we removed all hard references between the controllers, and relied solely on the controller messenger.